### ISSUE TITLE
Apex and other DC++ clients unable to get filelist 

### ISSUE THREAD

**Issue by drd33m** [aler9/dctoolkit#8]
Hi thankyou for this library it is one of the only modern NMDC/ADC client libs out there.

I am having an issue with getting the filelist from DC clients. When I try to get the filelist of the bot I am instanly disconnected from them and no filelist is downloaded.

I have done some testing and I can get filelist bot to bot though so I am unsure what is causing the issue
I have attached a log of my getting the filelist successfully another program using the same lib and me trying to get it with a DC client
```
$ConnectToMe UnCoolBarry xx:8989
2020/02/19 20:11:50 [peer] outgoing xx:8989
2020/02/19 20:11:50 [peer] connected xx:8989
$MyNick +Barry
$Lock EXTENDEDPROTOCOLEXTENDEDPROTOCOLABCABCABCABCABCABC Pk=DCPLUSPLUS0.868
$Supports Minislots XmlBZList ADCGet TTHL TTHF ZLIG
$Direction Download 22221
$Key A ѱ0A ѱ00 0 0 0 0 0
$ADCGET file files.xml.bz2 0 -1 ZL1
2020/02/19 20:11:51 [upload] [+Barry] request filelist (s=0 l=-1)
2020/02/19 20:11:51 [upload] [+Barry] finished filelist (s=0 l=1622)
2020/02/19 20:11:55 [peer] disconnected
$Quit +Barry
```

DC client (ApexDC++ 1.6.5)
```
$ConnectToMe UnCoolBarry xx:2020
2020/02/19 20:12:10 [peer] outgoing xx:2020
2020/02/19 20:12:10 [peer] connected xx:2020
$MyNick me\n
$Lock EXTENDEDPROTOCOLABCABCABCABCABCABC Pk=DCPLUSPLUS0.842
$Supports MiniSlots XmlBZList ADCGet TTHL TTHF ZLIG
$Direction Download 2720
$Key A ѱ0A ѱ00 0 0 0 0 0
$ADCGET file files.xml.bz2 0 -1 ZL1 DB0
2020/02/19 20:12:11 [peer] disconnected
```

**Comment by aler9** [comment]
Hello, thanks for the report, from what i read it seems that ApexDC inserts a strange flag "DB0" at the end of $ADCGET:
```
$ADCGET file files.xml.bz2 0 -1 ZL1 DB0
```

I've never seen this DB0 and this library doesn't handle it, as you can see from the ADCGET parser here:
https://github.com/direct-connect/go-dc/blob/master/nmdc/files.go#L93

I'll investigate this as soon as possible, and then i'll come up with a patch.

**Comment by aler9** [comment]
It seems that ApexDC++ always inserts a DB[downloaded bytes] string at the end of ADCGet:
```
AdcCommand Download::getCommand(bool zlib) const {
	AdcCommand cmd(AdcCommand::CMD_GET);
	
	cmd.addParam(Transfer::names[getType()]);

	if(getType() == TYPE_PARTIAL_LIST) { 
		cmd.addParam(Util::toAdcFile(getTempTarget()));
	} else if(getType() == TYPE_FULL_LIST) {
		if(isSet(Download::FLAG_XML_BZ_LIST)) {
			cmd.addParam(USER_LIST_NAME_BZ);
		} else {
			cmd.addParam(USER_LIST_NAME);
		}
	} else {
		cmd.addParam("TTH/" + getTTH().toBase32());
	}

	cmd.addParam(Util::toString(getStartPos()));
	cmd.addParam(Util::toString(getSize()));

	if(zlib && BOOLSETTING(COMPRESS_TRANSFERS)) {
		cmd.addParam("ZL1");
	}

	cmd.addParam("DB", Util::toString(downloadedBytes)); 

	return cmd;
}
```

Eiskalt DC++ has removed it:
https://github.com/eiskaltdcpp/eiskaltdcpp/blob/6c420aac3abcf59f3d3e24e90d3593a1b707720d/dcpp/Download.cpp#L104

SO: i'll edit go-dc and dctoolkit to support this flag, and then test it with ApexDC++.

**Comment by drd33m** [comment]
I did think it was the DB0 I tried removing it before it was paresed but I still had the same issue. I also tested it on DC++ client and the same thing happend

**Comment by aler9** [comment]
Hello, i think we're talking about at least 3 different problems that must be treated separately:

* the DB0 problem. I've added support for this field in go-dc and dctoolkit, and now when i create a dctoolkit-based bot i can download its filelist from ApexDC++ and DC++. This should be solved.

* the encryption problem. I was able to download the bot filelist from ApexDC++ only after disabling encryption. I've noticed that DC++ doesn't even support NMDC client-to-client encryption, while ApexDC++ tries to setup it, but it uses a very outdated openssl version, and thus it is kicked out by dctoolkit, that uses the standard golang tls library. The only solution to this consists in disabling encryption in dctoolkit, by initializing the client in this way:
```
client, err := dctk.NewClient(dctk.ClientConf{
		...
		PeerEncryptionMode: dctk.ForceEncryption,
	})
```

* the "More bytes written than requested" problem, this is not related to the filelist, and i think it should be treated in another issue with more details.
